### DISCUSSION TITLE
WebGPU Renderer

### DISCUSSION THREAD

**Discussion by obiot** [melonjs/melonJS#1030] [Ideas]
WebGPU is the working name for a future web standard and JavaScript API for accelerated graphics and compute, aiming to provide "modern 3D graphics and computation capabilities". It is developed in the W3C GPU for the Web Community Group with engineers from Apple, Mozilla, Microsoft, Google, and others.

few interesting readings :
https://hacks.mozilla.org/2020/04/experimental-webgpu-in-firefox/
https://webkit.org/demos/webgpu/
https://www.chromestatus.com/feature/6213121689518080

**Comment by L1lith** [comment]
[Babylon does a good job of advertising their webGPU compatibility on their site. ](https://www.babylonjs.com/) 3D game support would also be really cool for melon, I wonder if these two features could go hand in hand?

**Reply by joeldesmond60-byte** [reply]
When the balls land, the multipliers jump to the right, the multiplier bar is missing on all the mobile views, and the multiplier bar doesn’t fit properly on the out sizes

I needs to just fit these sizes Mobile S (320×568)
Mobile M (375×667)
Mobile L (425×812)
Popout L (800×450)
Popout S (400×225)
Laptop (1024×576)
Desktop (1200×675)

Can you handle that for me and what will you charge for that?
[PlinkoExtreme-1 (1).zip](https://github.com/user-attachments/files/22059648/PlinkoExtreme-1.1.zip)

**Comment by obiot** [comment]
definitely, same as WebGL today !

**Comment by obiot** [comment]
putting this here for later : https://github.com/gfx-rs/naga/

naga allows to convert GLSL to WSL, which would allows to automatically transpile the shader into what is required for WebGPU

**Reply by parasyte** [reply]
:+1:

I have been using `naga` for many years, and contributing to it on occasion!

**Comment by obiot** [comment]
we miss you on melonJS :)

**Comment by jonpchin** [comment]
I am really interested in webGPU support in melonJS. How much performance increase is upgrading WebGL 2 to use webGPU for a 2D game?

**Comment by obiot** [comment]
I honestly don't really expect a massive performance increase with WebGPU vs WebGL just for 2D rendering, but it's definitely a good way forward in terms of using a future-proof rendering method.

Are you interested like in willing to help starting a first implementation, or just generally interesting in having melonJS supporting WebGPU ?

**Reply by jonpchin** [reply]
I am generally interesting in having melonJS support WebGPU. WebGPU is still new/bleeding edge technology and not supported by all browsers yet. Do you have a time estimate of how long it would take someone to make a first implementation of WebGPU in melonJS? It may be better to wait for WebGPU to be more mature before starting a first implementation.

Edit: I am going to secretly try and take a stab at this first implementation on my free time. I am missing some expertise in this area but it will be something fun for me to do on my free time, I don't expect to get any pull requests merged into melonJS anytime soon.
